python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -0,0 +1,370 @@
+                ).records
+                other_materialized_assets = [
+                    event.event_log_entry.dagster_event.event_specific_data.asset_key
+                    for event in other_materialized_asset_records
+                ]
+                if current_asset in other_materialized_assets:
+                    # we still update the cursor for p so this materialization isn't considered
+                    # on the next sensor tick
+                    parent_asset_event_records[p] = (
+                        False,
+                        event_records[0].event_log_entry.timestamp,
+                    )
+                else:
+                    # current_asset was not updated along with p, so we consider p updated
+                    parent_asset_event_records[p] = (
+                        True,
+                        event_records[0].event_log_entry.timestamp,
+                    )
+            else:
+                # p has not been materialized and will not be materialized by the sensor
+                parent_asset_event_records[p] = (False, 0.0)
+
+    return parent_asset_event_records
+
+
+def _make_sensor(
+    selection: AssetSelection,
+    name: str,
+    wait_for_all_upstream: bool,
+    wait_for_in_progress_runs: bool,
+    minimum_interval_seconds: Optional[int] = None,
+    description: Optional[str] = None,
+    default_status: DefaultSensorStatus = DefaultSensorStatus.STOPPED,
+) -> MultiAssetSensorDefinition:
+    """Creates the sensor that will monitor the parents of all provided assets and determine
+    which assets should be materialized (ie their parents have been updated).
+
+    The cursor for this sensor is a dictionary mapping stringified AssetKeys to a timestamp (float). For each
+    asset we keep track of the timestamp of the most recent materialization of a parent asset. For example
+    if asset X has parents A, B, and C where A was materialized at time 1, B at time 2 and C at time 3. When
+    the sensor runs, the cursor for X will be set to 3. This way, the next time the sensor runs, we can ignore
+    the materializations prior to time 3. If asset A materialized again at time 4, we would know that this materialization
+    has not been incorporated into the child asset yet.
+    """
+
+    def sensor_fn(context):
+        asset_defs_by_key = (
+            context._repository_def._assets_defs_by_key  # pylint: disable=protected-access
+        )
+        source_asset_defs_by_key = (
+            context._repository_def.source_assets_by_key  # pylint: disable=protected-access
+        )
+        upstream: Mapping[AssetKey, Set[AssetKey]] = _get_upstream_mapping(
+            selection=selection,
+            assets=asset_defs_by_key.values(),
+            source_assets=source_asset_defs_by_key.values(),
+        )
+
+        cursor_dict: Dict[str, float] = json.loads(context.cursor) if context.cursor else {}
+        should_materialize: List[AssetKey] = []
```

Comment:
Maybe this should be a set because it's elements are supposed to be unique and we're often checking to see whether keys are included in it?